### ISSUE TITLE
index operation on a splat expression expands too early

### ISSUE THREAD

**Issue by splashx** [hashicorp/terraform#18051]
### Terraform Version
```
Terraform v0.11.7
+ provider.openstack v1.4.0
```

### Terraform Configuration Files
```hcl

resource "openstack_networking_port_v2" "port_network1" {
  name = "network1-${count.index+1}"
  count = "${var.vm_count}"
  admin_state_up = "true"
  network_id = "${var.vm_network1_id}"
  security_group_ids = ["${var.vm_sgs_network1}"]
}

resource "openstack_compute_instance_v2" "instance" {
  count = "${var.vm_count}"
  name = "some-name-${count.index+1}"
  image_id = "${var.vm_image_id}"
  flavor_name = "${var.vm_flavor_name}"
  key_pair = "${var.vm_keypair}"
  availability_zone = "${var.vm_availability_zone}"

  metadata = "${merge( var.vm_metadata, map("ansible_host", openstack_networking_port_v2.port_network1.*.all_fixed_ips.0[count.index] ))}"

  network {
    port = "${openstack_networking_port_v2.port_network1.*.id[count.index]}"
  }
  network {
    port = "${openstack_networking_port_v2.port_network2.*.id[count.index]}"
  }
  network {
    port = "${openstack_networking_port_v2.port_network3.*.id[count.index]}"
  }

}
```

### Debug Output

```
<.. clip ...>

2018/05/15 18:28:34 [DEBUG] Attaching resource state to "module.consul.openstack_networking_port_v2.port_network2[0]": &terraform.ResourceState{Type:"openstack_networking_port_v2", Dependencies:[]string{}, Primary:(*terraform.InstanceState)(0xc420bb1a90), Deposed:[]*terraform.InstanceState{}, Provider:"provider.openstack", mu:sync.Mutex{state:0, sema:0x0}}
2018/05/15 18:28:34 [DEBUG] Attaching resource state to "module.consul.openstack_networking_port_v2.port_network2[1]": &terraform.ResourceState{Type:"openstack_networking_port_v2", Dependencies:[]string{}, Primary:(*terraform.InstanceState)(0xc420bb1ae0), Deposed:[]*terraform.InstanceState{}, Provider:"provider.openstack", mu:sync.Mutex{state:0, sema:0x0}}
2018/05/15 18:28:34 [DEBUG] Attaching resource state to "module.consul.openstack_networking_port_v2.port_network2[2]": &terraform.ResourceState{Type:"openstack_networking_port_v2", Dependencies:[]string{}, Primary:(*terraform.InstanceState)(0xc420bb1bd0), Deposed:[]*terraform.InstanceState{}, Provider:"provider.openstack", mu:sync.Mutex{state:0, sema:0x0}}
2018/05/15 18:28:34 [DEBUG] Attaching resource state to "module.consul.openstack_networking_port_v2.port_network2[3]": &terraform.ResourceState{Type:"openstack_networking_port_v2", Dependencies:[]string{}, Primary:(*terraform.InstanceState)(0xc420bb1c20), Deposed:[]*terraform.InstanceState{}, Provider:"provider.openstack", mu:sync.Mutex{state:0, sema:0x0}}
2018/05/15 18:28:34 [DEBUG] Attaching resource state to "module.consul.openstack_networking_port_v2.port_network2[4]": &terraform.ResourceState{Type:"openstack_networking_port_v2", Dependencies:[]string{}, Primary:(*terraform.InstanceState)(0xc420bb1c70), Deposed:[]*terraform.InstanceState{}, Provider:"provider.openstack", mu:sync.Mutex{state:0, sema:0x0}}
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_networking_port_v2.port_network2[0]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_networking_port_v2.port_network2[1]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_networking_port_v2.port_network2[2]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_networking_port_v2.port_network2[3]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_networking_port_v2.port_network2[4]" references: []
2018-05-15T18:28:34.205Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.210Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.211Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.212Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.225Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.225Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.225Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.225Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.226Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.228Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.228Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018-05-15T18:28:34.235Z [DEBUG] plugin.terraform-provider-openstack_v1.4.0_x4: 2018/05/15 18:28:34 [DEBUG] Instance Diff is nil in Diff()
2018/05/15 18:28:34 [DEBUG] Resource state not found for "module.consul.openstack_compute_instance_v2.instance[2]": module.consul.openstack_compute_instance_v2.instance[2]
2018/05/15 18:28:34 [DEBUG] Resource state not found for "module.consul.openstack_compute_instance_v2.instance[3]": module.consul.openstack_compute_instance_v2.instance[3]
2018/05/15 18:28:34 [DEBUG] Resource state not found for "module.consul.openstack_compute_instance_v2.instance[4]": module.consul.openstack_compute_instance_v2.instance[4]
2018/05/15 18:28:34 [DEBUG] Resource state not found for "module.consul.openstack_compute_instance_v2.instance[0]": module.consul.openstack_compute_instance_v2.instance[0]
2018/05/15 18:28:34 [DEBUG] Resource state not found for "module.consul.openstack_compute_instance_v2.instance[1]": module.consul.openstack_compute_instance_v2.instance[1]
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_compute_instance_v2.instance[0]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_compute_instance_v2.instance[1]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_compute_instance_v2.instance[2]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_compute_instance_v2.instance[3]" references: []
2018/05/15 18:28:34 [DEBUG] ReferenceTransformer: "module.consul.openstack_compute_instance_v2.instance[4]" references: []
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalInterpolate, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalSequence, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalInterpolate, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalSequence, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalInterpolate, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalSequence, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalInterpolate, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalSequence, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalInterpolate, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [ERROR] root.consul: eval: *terraform.EvalSequence, err: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
2018/05/15 18:28:34 [DEBUG] plugin: waiting for all plugin processes to complete...

Error: Error running plan: 1 error(s) occurred:

* module.consul.openstack_compute_instance_v2.instance: 5 error(s) occurred:

* module.consul.openstack_compute_instance_v2.instance[4]: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
* module.consul.openstack_compute_instance_v2.instance[0]: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
* module.consul.openstack_compute_instance_v2.instance[1]: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
* module.consul.openstack_compute_instance_v2.instance[2]: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
* module.consul.openstack_compute_instance_v2.instance[3]: Resource 'openstack_networking_port_v2.port_network1' does not have attribute 'all_fixed_ips.0' for variable 'openstack_networking_port_v2.port_network1.*.all_fixed_ips.0'
```

### Expected Behavior

All resources should be applied.

### Actual Behavior

The computed attribute is not marked as dependent and `apply` fails.

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

### Additional Context

Originally the value of `count` was `2` and this line was not in the module:
```
 metadata = "${merge( var.vm_metadata, map("ansible_host", openstack_networking_port_v2.port_network1.*.all_fixed_ips.0[count.index] ))}"
```

After adding the above line and running `apply`, everything went okay.
After increasing the `count` to `5`, the error appears for the 3 new instances (this debug was run after a complete `destroy`, so it fails in all 5 instances, not only 3)

**Comment by jbardin** [comment]
Hi @splashx, 

Thanks for the interesting example here. I'm not sure exactly what's happening yet, but I have a feeling it's not really the resource dependency order since the graph looks correct, but more the order of evaluation in the overall interpolation expression.

I actually haven't reproduced this yet with the exact same error, but could you try using the `element` function rather than the index?

     metadata = "${merge(var.vm_metadata, map("ansible_host", element(openstack_networking_port_v2.port_network1.*.all_fixed_ips.0, count.index)))}"

**Comment by splashx** [comment]
Hi @jbardin. Thanks for jumping on this.

The outcome is different, but still an error:


```
Error: Error refreshing state: 1 error(s) occurred:

* module.consul.openstack_compute_instance_v2.instance: 1 error(s) occurred:

* module.consul.openstack_compute_instance_v2.instance[5]: index 5 out of range for list openstack_networking_port_v2.port_network1.*.all_fixed_ips.0 (max 5) in:

${merge( var.vm_metadata,
                      map("ansible_host", openstack_networking_port_v2.port_network1.*.all_fixed_ips.0[count.index] ),
                      map("keepalived_interface_ip", element(openstack_networking_port_v2.port_network2.*.all_fixed_ips.0, count.index))
                      )}
```

**Comment by jbardin** [comment]
It looks like you still have the same `ansible_host` line with the same index expression. 

If changing that to use `element` here doesn't work on it's own, you may be able to break up the expression into multiple stages by assigning the intermediate lists to local values.

**Comment by teamterraform** [comment]
Hi @splashx! Sorry for the long silence here.

Revisiting this bug some time later, it seems like another example of a common family of evaluation errors in Terraform 0.11 and prior. The other issues of this type were confirmed fixed in Terraform 0.12, and indeed the codepath that would generate the errors you saw here doesn't even exist anymore.

With that said, we're going to close this out now that Terraform 0.12 is released. If you or anyone else is still having trouble with quirky behavior in splat expressions, please open a new issue for it since the splat expression implementation is completely changed in Terraform 0.12 and so we'll need to capture a new set of debug logs, etc in order to analyze it.

Thanks for reporting this, and sorry again for the long silence.
